// agent-skills-eval/dist/evaluate-skills.js
import path from "node:path";
import { mkdirSync, rmSync, writeFileSync } from "node:fs";
import { allocateHistoryIteration, allocateIterationWorkspace, buildBenchmark, ensureSkillWorkspaceDir, snapshotSkillToHistory, } from "./artifacts.js";
import { consoleReporter } from "./console-reporter.js";
import { discoverSkills } from "./discover.js";
import { generateReport } from "./report.js";
import { runEval } from "./run-eval.js";
import { loadSkill } from "./skill.js";
import { slugify } from "./fs-utils.js";
/**
 * Tiny bounded-concurrency worker pool. Each worker grabs the next item off a
 * shared FIFO queue and runs `work` on it; resolves once the queue is drained.
 * Order of completion is non-deterministic across items.
 */
async function runPool(items, n, work) {
    const queue = items.slice();
    const workers = Array.from({ length: Math.max(1, Math.min(n, items.length)) }, async () => {
        while (queue.length > 0) {
            const next = queue.shift();
            if (next === undefined)
                return;
            await work(next);
        }
    });
    await Promise.all(workers);
}
export async function evaluateSkills(args) {
    const refs = discoverSkills(args.root, { include: args.include, exclude: args.exclude }).filter((ref) => ref.hasEvals);
    const modes = args.baseline ? ["with_skill", "without_skill"] : ["with_skill"];
    const workspaceLayout = args.workspaceLayout ?? "flat";
    const runWorkspace = workspaceLayout === "iteration"
        ? allocateIterationWorkspace(args.workspace)
        : { dir: args.workspace, iteration: undefined };
    // Default to the bundled rich reporter when the caller didn't wire any
    // observability — keeps the out-of-the-box CLI experience informative.
    const emit = args.onEvent
        ? args.onEvent
        : args.onLog
            ? undefined
            : consoleReporter();
    const concurrency = Math.max(1, args.concurrency ?? 4);
    // ─── Phase 1: sequential discovery prep ───────────────────────────────────
    // Allocate skillDir + write meta.json + emit suite-start in discovery order
    // so the start-of-run banner is stable run-to-run regardless of pool size.
    const prepared = [];
    for (const ref of refs) {
        args.onLog?.(`skill ${ref.name}: loading ${ref.relPath}`);
        const skill = loadSkill(ref.dir, { strict: args.strict });
        const slug = slugify(skill.name);
        const skillDir = workspaceLayout === "flat"
            ? ensureSkillWorkspaceDir(runWorkspace.dir, slug)
            : refs.length === 1
                ? prepareIterationSkillDir(runWorkspace.dir)
                : prepareIterationSkillDir(path.join(runWorkspace.dir, slug));
        writeFileSync(path.join(skillDir, "meta.json"), `${JSON.stringify({
            name: skill.name,
            slug,
            relPath: ref.relPath,
            target: args.target.model,
            judge: args.judge.model,
            modes,
            generated_at: new Date().toISOString(),
        }, null, 2)}\n`, "utf-8");
        emit?.({
            type: "suite-start",
            skill: skill.name,
            relPath: ref.relPath,
            evalsCount: skill.evals.length,
            modes,
            target: args.target.model,
            judge: args.judge.model,
        });
        prepared.push({ ref, skill, slug, skillDir, aggregateRuns: [], passed: 0, failed: 0, completed: 0 });
    }
    // ─── Phase 2: flatten (skill, evalCase) tasks + run via worker pool ───────
    // Cross-skill ordering note: with concurrency > 1, suite-end for skill A
    // may fire before suite-end for skill B even if A appears later in
    // discovery — completion order tracks the pool, not the discovery walk.
    // Per-eval ordering is preserved: eval-start strictly precedes eval-end for
    // the same case (enforced inside runEval).
    const tasks = [];
    for (const p of prepared) {
        if (p.skill.evals.length === 0) {
            // No evals — finalize immediately so we still emit a benchmark + suite-end.
            const benchmark = buildBenchmark([]);
            const benchmarkPath = path.join(p.skillDir, "benchmark.json");
            writeFileSync(benchmarkPath, `${JSON.stringify(benchmark, null, 2)}\n`, "utf-8");
            emit?.({ type: "suite-end", skill: p.skill.name, benchmarkPath, benchmark });
            continue;
        }
        for (let index = 0; index < p.skill.evals.length; index++) {
            tasks.push({ prepared: p, evalCase: p.skill.evals[index], index });
        }
    }
    await runPool(tasks, concurrency, async (task) => {
        const { prepared: p, evalCase, index } = task;
        args.onLog?.(`skill ${p.skill.name}: eval ${evalCase.name ?? evalCase.id ?? index + 1}`);
        const result = await runEval({
            skill: p.skill,
            eval: evalCase,
            index,
            modes,
            target: args.target,
            judge: args.judge,
            workspace: runWorkspace.dir,
            evalRootDir: p.skillDir,
            iteration: 0,
            targetParams: args.targetParams,
            judgeParams: args.judgeParams,
            onEvent: emit,
        });
        // JS is single-threaded between awaits; the stat updates below run
        // atomically relative to other workers, so no explicit lock is needed.
        for (const mode of modes) {
            const modeResult = result.modes[mode];
            if (!modeResult)
                continue;
            p.aggregateRuns.push({
                mode,
                passRate: modeResult.grading.summary.pass_rate,
                durationMs: modeResult.timing.duration_ms,
                tokens: modeResult.timing.total_tokens,
            });
        }
        const withSkill = result.modes.with_skill;
        if (withSkill) {
            p.passed += withSkill.grading.summary.passed;
            p.failed += withSkill.grading.summary.failed;
        }
        p.completed++;
        if (p.completed === p.skill.evals.length) {
            const benchmark = buildBenchmark(p.aggregateRuns);
            const benchmarkPath = path.join(p.skillDir, "benchmark.json");
            writeFileSync(benchmarkPath, `${JSON.stringify(benchmark, null, 2)}\n`, "utf-8");
            emit?.({
                type: "suite-end",
                skill: p.skill.name,
                benchmarkPath,
                benchmark,
            });
        }
    });
    // ─── Phase 3: aggregate result + history snapshot + HTML report ───────────
    let totalPassed = 0;
    let totalFailed = 0;
    const skills = [];
    const writtenSkillDirs = [];
    for (const p of prepared) {
        totalPassed += p.passed;
        totalFailed += p.failed;
        const total = p.passed + p.failed;
        skills.push({
            skill: p.skill.name,
            slug: p.slug,
            relPath: p.ref.relPath,
            evals: p.skill.evals.length,
            passRate: total === 0 ? 1 : p.passed / total,
            benchmarkPath: path.join(p.skillDir, "benchmark.json"),
        });
        writtenSkillDirs.push({ slug: p.slug, dir: p.skillDir });
    }
    // Loop mode: snapshot the freshly-written workspace into a history slot so
    // the developer can compare this run against earlier iterations.
    let historyIteration;
    if (args.loop && writtenSkillDirs.length > 0) {
        const history = allocateHistoryIteration(args.workspace);
        for (const { slug, dir } of writtenSkillDirs) {
            snapshotSkillToHistory(dir, history.dir, slug);
        }
        historyIteration = history.iteration;
    }
    // Render the static HTML report unless explicitly disabled. `report.ts`
    // reads everything from disk, so it works equally well against a freshly
    // finished run or against a `.history/iteration-N/` snapshot later.
    let reportPath;
    if (args.report !== false) {
        const result = generateReport({
            workspace: runWorkspace.dir,
            target: args.target.model,
            judge: args.judge.model,
            title: args.reportTitle,
            output: args.reportOutput,
        });
        reportPath = result.reportPath;
    }
    return {
        passed: totalPassed,
        failed: totalFailed,
        skills,
        historyIteration,
        iteration: runWorkspace.iteration,
        workspaceRoot: runWorkspace.dir,
        reportPath,
    };
}
function prepareIterationSkillDir(dir) {
    rmSync(dir, { recursive: true, force: true });
    mkdirSync(dir, { recursive: true });
    return dir;
}
