// agent-skills-eval/dist/fs-utils.js
import { existsSync, mkdirSync, readFileSync, readdirSync, statSync, writeFileSync } from "node:fs";
import path from "node:path";
export const DEFAULT_MAX_FILE_BYTES = 64 * 1024;
export function toPosixPath(value) {
    return value.split(path.sep).join("/");
}
export const pathToPosix = toPosixPath;
export const normalizePosix = toPosixPath;
export function ensureDir(dir) {
    mkdirSync(dir, { recursive: true });
}
export function isInsideDir(root, candidate) {
    const rootDir = path.resolve(root);
    const absolutePath = path.resolve(candidate);
    return absolutePath === rootDir || absolutePath.startsWith(rootDir + path.sep);
}
export function assertInside(root, candidate, label = "path") {
    if (!isInsideDir(root, candidate)) {
        throw new Error(`${label} escapes ${root}`);
    }
}
export function safeResolve(root, relativePath) {
    const rootDir = path.resolve(root);
    const absolutePath = path.resolve(rootDir, relativePath);
    if (absolutePath !== rootDir && !absolutePath.startsWith(rootDir + path.sep)) {
        return null;
    }
    return { absolutePath, relativePath: toPosixPath(path.relative(rootDir, absolutePath)) };
}
export function resolveInside(root, relativePath) {
    const resolved = safeResolve(root, relativePath);
    if (!resolved)
        throw new Error(`Path escapes skill directory: ${relativePath}`);
    return resolved.relativePath;
}
export function readAttachedFile(root, relativePath, maxFileBytes = DEFAULT_MAX_FILE_BYTES) {
    const resolved = safeResolve(root, relativePath);
    const normalized = resolved?.relativePath ?? toPosixPath(relativePath);
    if (!resolved || !existsSync(resolved.absolutePath)) {
        return { path: normalized, content: "", kind: "missing" };
    }
    const stat = statSync(resolved.absolutePath);
    if (!stat.isFile()) {
        return { path: normalized, content: "", kind: "missing" };
    }
    const buffer = readFileSync(resolved.absolutePath);
    const probe = buffer.subarray(0, Math.min(buffer.length, 4096));
    if (probe.includes(0)) {
        return { path: normalized, content: "", kind: "binary-skipped", bytes: stat.size };
    }
    if (buffer.length > maxFileBytes) {
        return {
            path: normalized,
            content: buffer.subarray(0, maxFileBytes).toString("utf-8"),
            kind: "too-large",
            bytes: stat.size,
        };
    }
    return { path: normalized, content: buffer.toString("utf-8"), kind: "text", bytes: stat.size };
}
export function writeFileInside(root, relativePath, content) {
    const resolved = safeResolve(root, relativePath);
    if (!resolved)
        throw new Error(`Output file escapes output directory: ${relativePath}`);
    ensureDir(path.dirname(resolved.absolutePath));
    writeFileSync(resolved.absolutePath, content);
}
export function safeReadJson(filePath) {
    if (!existsSync(filePath))
        return undefined;
    const parsed = JSON.parse(readFileSync(filePath, "utf-8"));
    return parsed && typeof parsed === "object" && !Array.isArray(parsed)
        ? parsed
        : undefined;
}
function escapeXml(value) {
    return value
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;");
}
export function attachedFileXml(tag, file) {
    const attrs = `path="${escapeXml(file.path)}" kind="${file.kind}"${file.bytes === undefined ? "" : ` bytes="${file.bytes}"`}`;
    if (file.kind === "text" || file.kind === "too-large") {
        return `<${tag} ${attrs}>\n${file.content}\n</${tag}>`;
    }
    return `<${tag} ${attrs}>${file.kind}</${tag}>`;
}
export function slugify(value, fallback = "item") {
    const slug = value.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "").slice(0, 64);
    return slug || fallback;
}
export function walkFiles(root, predicate) {
    if (!existsSync(root))
        return [];
    const out = [];
    const visit = (dir) => {
        for (const entry of readdirSync(dir, { withFileTypes: true })) {
            const absolutePath = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                visit(absolutePath);
            }
            else if (entry.isFile() && predicate(absolutePath)) {
                out.push(absolutePath);
            }
        }
    };
    visit(root);
    return out.sort((a, b) => a.localeCompare(b));
}
